core/src/typ.rs

```diff
@@ -269,6 +269,8 @@ pub enum TypeF<Ty, RRows, ERows> {
     ///
     /// See [`crate::term::Term::Sealed`].
     Symbol,
+    /// An opaque value, the type of `Term::Opaque`.
+    Opaque,
```

Comment:
I wonder if we should even worry about having a type for that, given that the user can't do much with opaque values, and can't use this type right now (it's not in the grammar, and it would be a breaking change to add it). However it doesn't cost much, and maybe it could be useful for Rust binaries consuming the library? I'm not sure.

Reply:
I was thinking that the nickel embedder might provide a function like `union: ForeignId -> ForeignId -> ForeignId`, so it would be nice if we could type-check it

Reply:
Ah, you mean customizing the set of operations possible on opaque values with their own operations? There is still the problem that if the user can't write this type down, it is somehow a second class citizen (and once again, making it possible to spell out is a breaking change).

All of that being said, having the type internally is like 10 additional line of code, so let's not argue about it too much. It doesn't really hurt to have this type internally, and see later if it can be turned into something useful. I'm fine with keeping it as it is for now

Reply:
A second idea to make it not a breaking change is to wait for the let-type RFC, and make it possible to define primitive types - think `%Opaque%` (just so it doesn't clash with user-land). Then, once we can export types from a record, we could define stdlib namespaced types, like `type Opaque = %Opaque%` in `std.foreign` or something.

Reply:
Perhaps a good way to do this would be to have a `newtype` construct to allow creating a distinct type family that is backed by an existing type, so that we can have this core Opaque/ForeignId type, then give names and possibly generics to various kinds of foreign data. That allows making more specific function signatures that work on foreign data too.

Reply:
Yeah, in fact the `let type` or `type = ` would precisely act like a `newtype`.  But the point stands that you need to be able to refer to a primitive `Opaque` type at some point, and just adding it would break backward compatibility (maybe someone called their contract `Opaque` already). So the idea would be to have the primitive type use an obscure internal syntax - as other primops work in nickel already - and the stdlib would just export `type Opaque = %BuiltinOpaqueType%`.